### DISCUSSION TITLE
Simulating Target-Type Inference in C#

### DISCUSSION THREAD

**Discussion by giscus** [Tyrrrz/Tyrrrzme#31] [Blog]
# Simulating Target-Type Inference in C#

Above everything else in software development, I really enjoy building frameworks that enable other developers to create something cool. Sometimes, when chasing that perfect design I have in mind, I find myself coming up with weird hacks that push the C#…

https://tyrrrz.me/blog/target-type-inference

**Comment by gtechsltn** [comment]
C# OOP: Greate Article! Thank you so much!

**Comment by cartercanedy** [comment]
You can make use of implicit conversion of the result target types directly by implementing conversion operators for both T & E to Result<T, E>, which removes the need for an intermediate/late-bound Result type. It's not necessarily explicit, but it feels better, imho.

**Reply by Tyrrrz** [reply]
Can you please show an example of what you mean?

**Reply by cartercanedy** [reply]
```csharp
// ...
public static implicit operator Result<T, E>(T ok) => new Ok(ok);
public static implicit operator Result<T, E>(E err) => new Err(err);
// ...
```

**Reply by jv-np** [reply]
This is usually fine, but I think there are problems with the approach.
One of them is that it starts to feel too much implicit, so you can get easily lost on what result should represent
Other is that you lose the ability to distinct between 'ok' and 'error' cases, like the last part of the article solves

I think a good middle ground is to allow the Ok case to be explicit, but for the error case you only allow types constrained by some 'IError' interface, and then allow those to be implicitly convertable to results.

**Reply by cartercanedy** [reply]
> This is usually fine, but I think there are problems with the approach.

The best part is that neither approach is mutually exclusive, I'm currently using both in my codebases :)

Just to be clear, I found this article extremely interesting and valuable. I also want to make the point that a lot of the utility of these data structures is the algorithms that they enable for chaining fallible operations easily (`map`, `and_then`, `or_else`, etc.). Both approaches are just the entry points
